src/ir/auth_logic/ast_construction.cc

```diff
@@ -0,0 +1,153 @@
+  return dynamic_cast<T*>(context);
+}
+
+static BaseFact ConstructVerbphrase(
+    Principal left_principal, AuthLogicParser::VerbphraseContext* context) {
+  CHECK(context != nullptr) << "Invalid Verbphrasecontext";
+  if (auto* ctx = GetVerbPhraseContextAs<AuthLogicParser::PredphraseContext>(
+          context)) {  // PredicateContext
+    return BaseFact(
+        Attribute(left_principal, ConstructPredicate(ctx->predicate())));
+  }
+  auto* ctx = GetVerbPhraseContextAs<AuthLogicParser::ActsAsPhraseContext>(
+      context);  // ActsAsPhraseContext
+  return BaseFact(
+      CanActAs(left_principal, ConstructPrincipal(ctx->principal())));
+}
+
+template <typename T, typename = std::enable_if_t<std::is_base_of<
+                          AuthLogicParser::FlatFactContext, T>::value>>
+T* GetFlatFactContextAs(AuthLogicParser::FlatFactContext* context) {
+  return dynamic_cast<T*>(context);
+}
+
+static BaseFact ConstructFlatFact(AuthLogicParser::FlatFactContext* context) {
+  CHECK(context != nullptr) << "Invalid FlatFactcontext";
+  if (auto* ctx =
+          GetFlatFactContextAs<AuthLogicParser::PrinFactContext>(context)) {
+    return ConstructVerbphrase(ConstructPrincipal(ctx->principal()),
+                               ctx->verbphrase());
+  }
+  auto* ctx = GetFlatFactContextAs<AuthLogicParser::PredFactContext>(
+      context);  // predFactContext
+  return BaseFact(ConstructPredicate(ctx->predicate()));
+}
+
+template <typename T, typename = std::enable_if_t<std::is_base_of<
+                          AuthLogicParser::FactContext, T>::value>>
+T* GetFactContextAs(AuthLogicParser::FactContext* context) {
+  return dynamic_cast<T*>(context);
+}
+
+static Fact ConstructFact(AuthLogicParser::FactContext* context) {
+  CHECK(context != nullptr) << "Invalid Factcontext";
+  if (auto* flat_fact_fact_ctx =
+          GetFactContextAs<AuthLogicParser::FlatFactFactContext>(
+              context)) {  // FlatFactFactContext
+    return Fact(ConstructFlatFact(flat_fact_fact_ctx->flatFact()));
+  }
+  auto* ctx = GetFactContextAs<AuthLogicParser::CanSayFactContext>(context);
+  return ConstructFact(ctx->fact());
+}
+
+static Query ConstructQuery(AuthLogicParser::QueryContext* context) {
+  return Query(context->ID()->getText(),
+               ConstructPrincipal(context->principal()),
+               ConstructFact(context->fact()));
+}
+
+static Program ConstructProgram(AuthLogicParser::ProgramContext* context) {
+  CHECK(context != nullptr) << "Invalid programcontext";
+  std::vector<SaysAssertion> says_assertions;
+  std::vector<Query> queries;
+
+  // for ( auto const &statement : context->saysAssertion()) {
+  //     says_assertions_.push_back(std::move(construct_says_assertion(statement)));
+  // }
+
+  for (auto const& statement : context->query()) {
+    CHECK(statement != nullptr) << "Invalid Querycontext";
+    Query query = ConstructQuery(statement);
+    queries.push_back(std::move(query));
+  }
```

Comment:
What is the return type of `context->query()?

```suggestion
std::vector<Query> queries = utils::MapIter<Query>(context->query(), [](const auto* statement) {
    return ConstructQuery(CHECK_NOTNULL(statement));
}
```